### DISCUSSION TITLE
About Velox's Spiller

### DISCUSSION THREAD

**Discussion by waitinfuture** [facebookincubator/velox#5021] [Ideas]
Seems some behavior of Velox's Spiller is not very efficient:
1. Each time ```Spiller::spill``` is called, ```fillSpillRuns``` will be called to calculate the hash value for every row in RowContainer, but spill runs(along with the partitionId for each row) which are not selected will be cleared. This will possibly cause re-calculation of the hash values in later calls of ```Spiller::spill```.
2. At least one partition of rows will be spilled, possibly much larger than the required memory.

I'm thinking whether it's possible to encode the hash value for each row when type is ```kAggregate``` and ```kHashJoinBuild``` inside the row layout, with some bookkeeping we can avoid those problems. The encoded hash value can still be calculated in batch. And since memory arbitrator plans to do compaction in the future, it's possible to use the hash value to compact rows of the same partition together.

cc @xiaoxmeng @mbasmanova

**Comment by mbasmanova** [comment]
@xiaoxmeng Meng, would you take a look?
